In [1]:
# import necessary libraries
import pandas as pd
import numpy as np
import seaborn as sns
sns.set()
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')


In [2]:
# loaded the original dataset
df = pd.read_csv('imdb.csv')
df.tail(5)

Movie_Name  Release_Year Duration  Ratings Number_of_Ratings  \
9995             NaN          2023      NaN      7.2               668   
9996       Mafia Inc          2019      NaN      6.8              1.7K   
9997  Pound of Flesh          2015   1h 44m      5.1              7.2K   
9998             NaN          2014   1h 50m      6.0               445   
9999          Pledge          2018   1h 17m      5.4              3.9K   

                         Genres    Rated  Metascore         Director  \
9995        Comedy, Documentary   1h 14m        NaN  Manny Rodriguez   
9996                        NaN   2h 23m       64.0              NaN   
9997           Action, Thriller        R       41.0  Ernie Barbarash   
9998       Drama, Family, Music       PG        NaN       Andy Lauer   
9999  Horror, Mystery, Thriller  Unrated       54.0   Daniel Robbins   

                                                 Actors User Reviews  \
9995                                                NaN           16   
9996  Sergio Castellitto, Marc-André Grondin, Gilber...           24   
9997    Jean-Claude Van Damme, John Ralston, Aki Aleong           51   
9998                    David Kauffman, Steven Sessions            6   
9999   Zachery Byrd, Phillip Andre Botello, Zack Weiner           64   

     Critic_Reviews                                        Description  \
9995              4  Morgan showcases her life via intimate tales a...   
9996             13                                                NaN   
9997             44  A man's heroic attempt to help a woman in dist...   
9998              3  A singer and songwriter, who gave up his dream...   
9999             66  A group of college freshmen pledge an exclusiv...   

                                            Movie_Image  
9995  https://m.media-amazon.com/images/M/MV5BNmVkMz...  
9996  https://m.media-amazon.com/images/M/MV5BNTRkZj...  
9997  https://m.media-amazon.com/images/M/MV5BMjI2OD...  
9998  https://m.media-amazon.com/images/M/MV5BMGY5MT...  
9999  https://m.media-amazon.com/images/M/MV5BY2Q4OD...

In [3]:
# loaded the imputation dataset to impute it in original dataset missing values
data = pd.read_csv('imputation_imdb.csv')
data.tail()

Movie Name  Release Year Duration  Ratings  \
9995   Leanne Morgan: I'm Every Woman          2023   1h 14m      7.2   
9996                        Mafia Inc          2019   2h 23m      6.8   
9997                   Pound of Flesh          2015   1h 44m      5.1   
9998          The One I Wrote for You          2014   1h 50m      6.0   
9999                           Pledge          2018   1h 17m      5.4   

      Number_of_Ratings    Rated  Metascore  \
9995              668.0      NaN        NaN   
9996             1700.0      NaN       64.0   
9997             7200.0        R       41.0   
9998              445.0       PG        NaN   
9999             3900.0  Unrated       54.0   

                                            Description  \
9995  Morgan showcases her life via intimate tales a...   
9996  The Gamache, tailors from father to son, have ...   
9997  A man's heroic attempt to help a woman in dist...   
9998  A singer and songwriter, who gave up his dream...   
9999  A group of college freshmen pledge an exclusiv...   

                                            Movie Image  
9995  https://m.media-amazon.com/images/M/MV5BNmVkMz...  
9996  https://m.media-amazon.com/images/M/MV5BNTRkZj...  
9997  https://m.media-amazon.com/images/M/MV5BMjI2OD...  
9998  https://m.media-amazon.com/images/M/MV5BMGY5MT...  
9999  https://m.media-amazon.com/images/M/MV5BY2Q4OD...

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Movie Name         10000 non-null  object 
 1   Release Year       10000 non-null  int64  
 2   Duration           9875 non-null   object 
 3   Ratings            9970 non-null   float64
 4   Number_of_Ratings  9970 non-null   float64
 5   Rated              6402 non-null   object 
 6   Metascore          4921 non-null   float64
 7   Description        9977 non-null   object 
 8   Movie Image        9996 non-null   object 
dtypes: float64(3), int64(1), object(5)
memory usage: 703.3+ KB


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Movie_Name         8594 non-null   object 
 1   Release_Year       10000 non-null  int64  
 2   Duration           6719 non-null   object 
 3   Ratings            9970 non-null   float64
 4   Number_of_Ratings  9970 non-null   object 
 5   Genres             8284 non-null   object 
 6   Rated              9875 non-null   object 
 7   Metascore          4920 non-null   float64
 8   Director           8286 non-null   object 
 9   Actors             9913 non-null   object 
 10  User Reviews       9736 non-null   object 
 11  Critic_Reviews     9077 non-null   object 
 12  Description        8283 non-null   object 
 13  Movie_Image        10000 non-null  object 
dtypes: float64(2), int64(1), object(11)
memory usage: 1.1+ MB


## Data Preprocessing and Cleaning

In [6]:
# here can see the missing values in the Movie name column in the original dataset
df[df.Movie_Name.isnull()]

Movie_Name  Release_Year Duration  Ratings Number_of_Ratings  \
44          NaN          2024    3h 1m      6.7               37K   
47          NaN          2019   2h 41m      7.6              889K   
49          NaN          2024   2h 28m      7.5              264K   
59          NaN          2022   1h 47m      7.0              166K   
67          NaN          2024      NaN      7.6               27K   
...         ...           ...      ...      ...               ...   
9987        NaN          2016   1h 59m      7.0              1.3K   
9989        NaN          2022      NaN      6.1               426   
9990        NaN          2024      NaN      6.2               242   
9995        NaN          2023      NaN      7.2               668   
9998        NaN          2014   1h 50m      6.0               445   

                                                 Genres      Rated  Metascore  \
44                   Epic, Western Epic, Drama, Western          R       49.0   
47           Period Drama, Showbiz Drama, Comedy, Drama          A       84.0   
49    Action Epic, Car Action, Desert Adventure, Dys...          R       79.0   
59    Buddy Comedy, Satire, Spy, Action, Comedy, Cri...          A       68.0   
67    Period Drama, Swashbuckler, Action, Adventure,...     2h 58m       75.0   
...                                                 ...        ...        ...   
9987                                     Drama, Romance  Not Rated        NaN   
9989                    Crime, Drama, Mystery, Thriller        NaN        NaN   
9990                                     Drama, Mystery     1h 39m        NaN   
9995                                Comedy, Documentary     1h 14m        NaN   
9998                               Drama, Family, Music         PG        NaN   

                        Director  \
44                 Kevin Costner   
47             Quentin Tarantino   
49                 George Miller   
59                  Tom Gormican   
67    Alexandre de La Patellière   
...                          ...   
9987                  Jon Garcia   
9989                  Sean Olson   
9990              Austin Andrews   
9995             Manny Rodriguez   
9998                  Andy Lauer   

                                                 Actors User Reviews  \
44                Jon Baird, Kevin Costner, Mark Kasdan          687   
47          Leonardo DiCaprio, Brad Pitt, Margot Robbie         5.8K   
49          Anya Taylor-Joy, Chris Hemsworth, Tom Burke         1.1K   
59          Nicolas Cage, Pedro Pascal, Tiffany Haddish          777   
67    Alexandre Dumas, Alexandre de La Patellière, M...          130   
...                                                 ...          ...   
9987  Nick Ferrucci, Benjamin Farmer, Curtis Edward ...           18   
9989    Christy Barritt, D.F.W. Buckingham, Bryar Freed           13   
9990         Austin Andrews, J.M. Barrie, Andrew Holmes            8   
9995                                                NaN           16   
9998                    David Kauffman, Steven Sessions            6   

     Critic_Reviews                                        Description  \
44              136  Chronicles a multi-faceted, 15-year span of pr...   
47              678  As Hollywood's Golden Age is winding down duri...   
49              372  The origin story of renegade warrior Furiosa b...   
59              290  Moviestar Nick Cage is channeling his iconic c...   
67               94  After escaping from an island prison where he ...   
...             ...                                                ...   
9987              4  This third and final film of the Falls trilogy...   
9989              1  Gabby, a crime scene cleaner, finds overlooked...   
9990            NaN  A young woman follows a mysterious melody onto...   
9995              4  Morgan showcases her life via intimate tales a...   
9998              3  A singer and songwriter, who gave up his dream...   

                

In [7]:
# here i have extracted those movie names from imputation dataset where movie name column 
# of original dataset has null value
impute_movie_name = data.loc[df[df.Movie_Name.isnull()].index, 'Movie Name']

In [8]:
# and i assigned movie names to the original dataset where movie name column has missing values
df.loc[df.Movie_Name.isnull(), 'Movie_Name'] = impute_movie_name

In [11]:
# apparently many movies that aren't rated their duration values went to rated column
df[df['Duration'].isnull()].sample(3)

Movie_Name  Release_Year  \
7086                              We Don't Deserve Dogs          2020   
9781                          Kulej. Dwie strony medalu          2024   
2833   The Feeling That the Time for Doing Something...          2023   

     Duration  Ratings Number_of_Ratings       Genres   Rated  Metascore  \
7086      NaN      6.8               206  Documentary  1h 24m        NaN   
9781      NaN      6.5               234          NaN  2h 28m        NaN   
2833      NaN      5.7              1.3K       Comedy  1h 27m       77.0   

            Director                                          Actors  \
7086  Matthew Salleh                             3 nominations total   
9781             NaN  Tomasz Wlosok, Michalina Olszanska, Tomasz Kot   
2833    Joanna Arnow          Scott Cohen, Babak Tafti, Joanna Arnow   

     User Reviews Critic_Reviews  \
7086           11              7   
9781            1              2   
2833           10             40   

                                            Description  \
7086  A contemplative odyssey across our planet, loo...   
9781                                                NaN   
2833  A mosaic-style comedy following the life of a ...   

                                            Movie_Image  
7086  https://m.media-amazon.com/images/M/MV5BODU5MT...  
9781  https://m.media-amazon.com/images/M/MV5BNzg4Yz...  
2833  https://m.media-amazon.com/images/M/MV5BODIwMG...

In [12]:
# I'll get the duration values from rated column 
duration_in_rating_column = df[df.Duration.isnull()]['Rated']
# and impute null values instead
df.loc[df.Duration.isnull(), 'Rated'] = np.nan


In [14]:
# here i inserted the duration values from Rated column to Duration column
df.loc[df.Duration.isnull(),'Duration'] = duration_in_rating_column


In [17]:
# here i have extracted those Description from imputation dataset where Description column 
# of original dataset has null value
impute_description = data.loc[df[df.Description.isnull()].index, 'Description']


In [18]:
# and i assigned Description to the original dataset where Description column has missing values
df.loc[df.Description.isnull(), 'Description'] = impute_description


In [21]:
# unique values of rated column except null values
unique_rated = df['Rated'].unique()
unique_rated = unique_rated[~pd.isna(unique_rated)]


In [22]:
# I noticed some discrepancies, in dataset we have some rated values in duration
df[df['Duration'].isin(unique_rated)]


Movie_Name  Release_Year   Duration  \
1188                                 The Post          2023          R   
2046                       The Little Mermaid          2024          R   
4720                Lost & Found in Cleveland          2024      PG-13   
6936                   How to Make a Werewolf          2024        16+   
7664   Big City Greens the Movie: Spacecation          2024      TV-Y7   
8365                   OnlyFans: Selling Sexy          2021      TV-MA   
9536                       PMC: Deo beong-keo          2018      TV-14   
9747                           Sloppy Seconds          2023      TV-MA   
9794    Pater Noster and the Mission of Light          2024  Not Rated   

      Ratings Number_of_Ratings  \
1188      5.2               124   
2046      5.0                77   
4720      8.9                 9   
6936      6.0                75   
7664      6.2               504   
8365      4.9               172   
9536      5.4              1.6K   
9747      5.5                71   
9794      8.3               102   

                                                 Genres Rated  Metascore  \
1188                                              Drama   NaN        NaN   
2046                                   B-Horror, Horror   NaN        NaN   
4720                                                NaN   NaN        NaN   
6936                                                NaN   NaN        NaN   
7664  Space Sci-Fi, Action, Adventure, Animation, Co...   NaN        NaN   
8365                                                NaN   NaN        NaN   
9536                                             Action   NaN        NaN   
9747                                                NaN   NaN        NaN   
9794                                                NaN   NaN        NaN   

           Director                                             Actors  \
1188    Justin Hunt    Zachariah West, Angus Benfield, Kimberly Bigsby   
2046    Leigh Scott             Mike Markoff, Lydia Helen, Jeff Denton   
4720            NaN               Martin Sheen, June Squibb, Liza Weil   
6936            NaN       Alexis Andrews, Meghan Oliver, Jasmine Davis   
7664   Anna O'Brian         Chris Houghton, Shane Houghton, Lacey Dyer   
8365            NaN        Brittany Adeline, Camilla Araújo, Ana Cheri   
9536  Byung-woo Kim           Ha Jung-woo, Lee Sun-kyun, Jennifer Ehle   
9747            NaN  James Abernathy, Daniel Fernando Acosta, Gmac ...   
9794            NaN         Adara Starr, Mike Amason, Joshua R. Outzen   

     User Reviews Critic_Reviews  \
1188            8              3   
2046            2              2   
4720            5            NaN   
6936            7              9   
7664           11              5   
8365            1              1   
9536           31             15   
9747            5            NaN   
9794           11             29   

                                            Description  \
1188  An edgy, revenge thriller steeped in the undet...   
2046  Dr. Eric Prince, an archaeologist, makes a dra...   
4720  A 24-hour slice-of-life following five people ...   
6936  Urban action horror film that dives deep into ...   
7664  After ruining his family's vacation before it ...   
8365  Examines the voyeuristic and profit-making wor...   
9536  On the day of the U.S. presidential election i...   
9747  Keisha has been sheltered her entire life, but...   
9794  Max, a young record store clerk stumbles upon ...   

                                            Movie_Image  
1188  https://m.media-amazon.com/images/M/MV5BZDIwMz...  
2046  https://m.media-amazon.com/images/M/MV5BZjQyNW...  
4720  https://m.media-amazon.com/images/M/MV5BZjhiOD...  
6936  https://m.media-amazon.com/images/M/MV5BMGYwMj...  
7664  https://m.media-amazon.com/images/M/MV5BZjhkOG...  
8365  https://m.media-amazon.com/images/M/MV5BZmQ4Zj...  
9536  https://m.media-amazon.com/images/M/MV5BNzgzMj...  
9747  https://m.media-ama

In [23]:
# here i extracted the Rated values from duration column
rated_in_duration = df.loc[df['Duration'].isin(unique_rated), "Duration"]
# and assigned them to the right place in Rated column
df.loc[df['Duration'].isin(unique_rated), "Rated"] = rated_in_duration
# and for those duration assigned null values
df.loc[df['Duration'].isin(unique_rated), "Duration"] = np.nan


In [24]:
# now we have no such discrepancies in the dataset
df[df['Duration'].isin(unique_rated)]

Empty DataFrame
Columns: [Movie_Name, Release_Year, Duration, Ratings, Number_of_Ratings, Genres, Rated, Metascore, Director, Actors, User Reviews, Critic_Reviews, Description, Movie_Image]
Index: []

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Movie_Name         10000 non-null  object 
 1   Release_Year       10000 non-null  int64  
 2   Duration           9866 non-null   object 
 3   Ratings            9970 non-null   float64
 4   Number_of_Ratings  9970 non-null   object 
 5   Genres             8284 non-null   object 
 6   Rated              6728 non-null   object 
 7   Metascore          4920 non-null   float64
 8   Director           8286 non-null   object 
 9   Actors             9913 non-null   object 
 10  User Reviews       9736 non-null   object 
 11  Critic_Reviews     9077 non-null   object 
 12  Description        9977 non-null   object 
 13  Movie_Image        10000 non-null  object 
dtypes: float64(2), int64(1), object(11)
memory usage: 1.1+ MB


In [26]:
# here i have created a function to convert the duration values in numerical format for further analysis
# and to draw insights out of them
def duration_in_min(col):
    if pd.isna(col):
        return None
        
    col = str(col)
    
    if len(col.split(' ')) == 1:
        if 'm' in col:
            return int(col.replace('m', ''))
        elif 'h' in col:
            return int(col.replace('h', '')) * 60
            
    hour, minute = col.split(' ')
    
    if 'h' in hour and 'm' in minute:
        return int(hour.replace('h', ''))*60 + int(minute.replace('m', ''))
        
duration_in_min('2h')    

120

In [27]:
df['Duration(min)'] = df['Duration'].apply(duration_in_min)

In [28]:
# here i have created a function to convert the Number of Ratings values in numerical format for further analysis
# and to draw insights out of them
def rating_in_num(col):
    col = str(col)
    if 'K' in col:
        return float(col.replace('K', '')) * 1000     
    elif 'M' in col:
        return float(col.replace('M', '')) * 1000000
    else:    
        return float(col)

In [29]:

df['Number_of_Ratings(num)'] = df['Number_of_Ratings'].apply(rating_in_num)

In [31]:
# here i have created a function to convert the User Reviews values in numerical format for further analysis
# and to draw insights out of them
def user_review_num(col):
    col = str(col)
    if 'K' in col:
        return float(col.replace('K', '')) * 1000     
    else:    
        return float(col)

In [32]:
df['User Reviews(num)'] = df['User Reviews'].apply(user_review_num)

In [33]:
# here i have created a function to convert the Critic Reviews values in numerical format for further analysis
# and to draw insights out of them
def critic_review_num(col):
    col = str(col)
    if 'K' in col:
        return float(col.replace('K', '')) * 1000     
    else:    
        return float(col)

In [34]:
df['Critic_Reviews(num)'] = df['Critic_Reviews'].apply(critic_review_num)

In [36]:
'''a single movie has multiple genres, so i created a function that extract the Main genre from the movie
it will analyze a list of movie genres and match them to a predefined list of genres to find the most frequent one'''

from collections import Counter

# predefined genres that i got from imdb website
genres = ["Action", "Adventure", "Fantasy", "Drama", "Romance", 'Animation','Sci-Fi',
          'Crime', 'Comedy', 'Biography', 'Science Fiction', 'Horror', 'War', 
          'Family', 'Mystery', 'Thriller', 'Documentary', 'Musical', 'Sport', 
          'Western', 'Talk-Show', 'History', 'Game-Show', 'Music', 'News', 'Reality-TV']

def get_main_genre(movie_genres):
    
    if pd.isna(movie_genres):
        return None
    # Convert movie genres to lowercase and split by commas
    movie_genres = [genre.strip().lower() for genre in movie_genres.split(',')]
    
    # counter to keep track of matching genres
    genre_counts = Counter()
    
    # Iterate over each genre in the predefined genres list
    for predefined_genre in genres:
        normalized_predefined_genre = predefined_genre.lower()
        
        # Check how many times the predefined genre (or part of it) appears in movie genres
        for genre in movie_genres:
            if normalized_predefined_genre in genre:  # Match if part of the genre is in the movie's genre
                genre_counts[predefined_genre] += 1  # Increment the count for this predefined genre
    
    # Return the genre that matches the most
    if genre_counts:
        return genre_counts.most_common(1)[0][0]  # Get the genre with the highest count
    return None  # In case no genre matches

# Apply the function to the 'Genres' column to extract the main genre
df["Main_Genre"] = df['Genres'].apply(get_main_genre)


In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Movie_Name              10000 non-null  object 
 1   Release_Year            10000 non-null  int64  
 2   Duration                9866 non-null   object 
 3   Ratings                 9970 non-null   float64
 4   Number_of_Ratings       9970 non-null   object 
 5   Genres                  8284 non-null   object 
 6   Rated                   6728 non-null   object 
 7   Metascore               4920 non-null   float64
 8   Director                8286 non-null   object 
 9   Actors                  9913 non-null   object 
 10  User Reviews            9736 non-null   object 
 11  Critic_Reviews          9077 non-null   object 
 12  Description             9977 non-null   object 
 13  Movie_Image             10000 non-null  object 
 14  Duration(min)           9866 non-null  

## Modelling

In [43]:
# we have null values in Genre we can predict movie genres based on descriptions
df.loc[df.Main_Genre.isnull(), ['Main_Genre', 'Description']]

Main_Genre                                        Description
56         None  It follows an asteroid miner who, after crash-...
86         None  Murphy is an American living in Paris who ente...
129        None  Massimo is a member of the Sicilian Mafia fami...
142        None  Follows a group of people fueled by a profound...
223        None  A bright but socially awkward exchange student...
...         ...                                                ...
9977       None  Story of Vanessa Del Rio, the legendary adult ...
9981       None  It tells the story of Janice, the mother of a ...
9986       None  A retired all female heist crew is reactivated...
9994       None  After hunting a wolf cub, Veysi must deal with...
9996       None  The Gamache, tailors from father to son, have ...

[1717 rows x 2 columns]

In [45]:
# first I filled few missing values of description with movie name
movie_name_description = df[df.Description.isnull()]['Movie_Name']
df.loc[df['Description'].isnull(),  'Description'] = movie_name_description
df['Description'].isnull().sum()


0

__TF-IDF (Term Frequency-Inverse Document Frequency)__
* The TF-IDF vectorization process converts the movie descriptions (which are raw text data) into numerical representations that capture the importance of words in the context of the entire dataset.

* vectorizer.fit_transform takes your movie descriptions ('Description' column) and transforms them into a sparse matrix of numerical values, where each row corresponds to a movie, and each column corresponds to a unique word (term) in the entire dataset.
* The numerical value for each word represents its TF-IDF score for that particular movie description.

In [ ]:
# imported necessary libraries to create a ML model
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
# from xgboost import XGBClassifier


In [81]:

# Vectorize the descriptions using TF-IDF
vectorizer = TfidfVectorizer(stop_words='english')

# here the Description (where null values of genre are absent), are considered as a feature.
X = vectorizer.fit_transform(df.loc[~df.Main_Genre.isnull(), 'Description'])
# genre is the target that I am trying to predict based on the description
y = df.loc[~df.Main_Genre.isnull(), 'Main_Genre']


In [82]:
# split the data into training and testing set
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=1)


In [83]:

# here i have trained the RandomForestClassifier model on training data
classifier = RandomForestClassifier()
classifier.fit(x_train, y_train)


RandomForestClassifier()

In [84]:
# performance on testing set
predicted = classifier.predict(x_test)
accuracy_score(y_test, predicted)


0.4571773220747889

We're predicting genres in a dataset with 21 different genre classes, the baseline accuracy (just guessing one class out of 21) would be 4.7%. So 45.7% would be considered much better than random guessing.

In [85]:
# performance on training set
predicted = classifier.predict(x_train)
accuracy_score(y_train, predicted)


1.0

In [92]:
# now i have trained the model on whole data
classifier = RandomForestClassifier()
classifier.fit(X, y)


RandomForestClassifier()

In [93]:
# i use the model to predict genres for the movie descriptions that do not have a genre
description_for_prediction = vectorizer.transform(df.loc[df.Main_Genre.isnull(), 'Description'])
predicted_genres = classifier.predict(description_for_prediction)
predicted_genres


array(['Horror', 'Drama', 'Drama', ..., 'Action', 'Drama', 'Drama'],
      dtype=object)

In [95]:
# here i inserted the predicted genres back to dataframe where Main_genre had missing values
df.loc[df.Main_Genre.isnull(), 'Main_Genre'] = predicted_genres


In [96]:
df.isnull().sum()


Movie_Name                   0
Release_Year                 0
Duration                   134
Ratings                     30
Number_of_Ratings           30
Genres                    1716
Rated                     3272
Metascore                 5080
Director                  1714
Actors                      87
User Reviews               264
Critic_Reviews             923
Description                  0
Movie_Image                  0
Duration(min)              134
Number_of_Ratings(num)      30
User Reviews(num)          264
Critic_Reviews(num)        923
Main_Genre                   0
dtype: int64

In [106]:
# it would be wise to add Main genre value to fill missing genres in genres column
df.loc[df.Genres.isnull(), 'Genres'] = df.loc[df.Genres.isnull(), 'Main_Genre'] 


In [108]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Movie_Name              10000 non-null  object 
 1   Release_Year            10000 non-null  int64  
 2   Duration                9866 non-null   object 
 3   Ratings                 9970 non-null   float64
 4   Number_of_Ratings       9970 non-null   object 
 5   Genres                  10000 non-null  object 
 6   Rated                   6728 non-null   object 
 7   Metascore               4920 non-null   float64
 8   Director                8286 non-null   object 
 9   Actors                  9913 non-null   object 
 10  User Reviews            9736 non-null   object 
 11  Critic_Reviews          9077 non-null   object 
 12  Description             10000 non-null  object 
 13  Movie_Image             10000 non-null  object 
 14  Duration(min)           9866 non-null  

## Data Ingestion into MySQL

In [ ]:
# imported necessary libraries to load the cleaned prepared dataframe into MySQL 
from sqlalchemy import create_engine
import mysql.connector
import datetime


In [109]:

start_time = datetime.datetime.now()
print('Begin:', start_time)

engine = create_engine('mysql+pymysql://root:Rafi%40mysql123@localhost:3306/imdb?charset=utf8mb4')

try:
    df.to_sql(name='imdb_10k_movies', con=engine, index=False, if_exists='replace')
    print("Successfully imported")
    
except Exception as e:
    print(f"Failed to import: {i}. Error: {e}")


end_time = datetime.datetime.now()
print('End:', end_time)

total_time = end_time - start_time
print('Total time:', total_time)


Begin: 2025-02-01 23:24:32.639526
Successfully imported
End: 2025-02-01 23:24:34.216764
Total time: 0:00:01.577238
